File: app/android/app/src/main/kotlin/com/sheyaln/aac/CloudBackup.kt
package com.sheyaln.aac

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.provider.DocumentsContract
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.Executors

/**
 * Copies of the board in a folder the caregiver picked, in their own account.
 *
 * The folder is chosen once through the system document picker, so it can be
 * their Google Drive, their OneDrive, or the tablet's own storage — whichever
 * providers that device has. What comes back is a permission scoped to that one
 * folder, held across reboots, and every copy afterwards is automatic.
 *
 * **This app holds no Google credential and never asks for one.** The rejected
 * alternative was Drive's `appDataFolder` through the Drive REST API, which is
 * the closer match to the iOS side's hidden container: it needs an OAuth client
 * tied to the upload key and Google's verification of the `drive.appdata`
 * sensitive scope. Both are infrastructure on the developer's side of the line,
 * for a feature whose whole claim is that there is no developer side.
 *
 * Android's own Auto Backup was rejected for the reason the iCloud device
 * backup was: silent, capped at 25 MB, impossible to list or date, and it comes
 * back only on a reinstall. A backup a caregiver cannot verify is the failure
 * this feature exists against, not a cheaper version of the fix.
 *
 * Every call runs off the main thread. A copy of a board is megabytes over a
 * content provider that may be a network away, and a blocked main thread here
 * is a nonspeaking person holding a frozen board.
 */
class CloudBackup(private val activity: Activity, messenger: BinaryMessenger) {
  private val channel = MethodChannel(messenger, "org.wordbridge/cloud_backup")
  private val work = Executors.newSingleThreadExecutor()
  private val main = Handler(Looper.getMainLooper())

  /** The picker this is waiting on, or null. Only one can be open at a time. */
  private var pending: MethodChannel.Result? = null

  init {
    channel.setMethodCallHandler { call, result ->
      when (call.method) {
        // Answered here rather than on the worker: it opens a picker in front
        // of whoever is holding the tablet, so it belongs on the thread that
        // owns the screen.
        "connect" -> connect(call, result)
        else ->
          work.execute {
            try {
              val value =
                when (call.method) {
                  "place" -> where()
                  "list" -> list()
                  "upload" ->
                    upload(call.argument<String>("path")!!, call.argument<String>("name")!!)
                  "download" -> {
                    download(call.argument<String>("id")!!, call.argument<String>("path")!!)
                    null
                  }
                  "delete" -> {
                    delete(call.argument<String>("id")!!)
                    null
                  }
                  else -> {
                    main.post { result.notImplemented() }
                    return@execute
                  }
                }
              main.post { result.success(value) }
            } catch (e: Refusal) {
              main.post { result.error(e.code, e.detail, null) }
            } catch (e: Exception) {
              main.post { result.error("failed", e.toString(), null) }
            }
          }
      }
    }
  }

  /**
   * A reason the Dart side turns into a sentence.
   *
   * The strings here never reach a caregiver — `refusalFor` in
   * cloud_destination.dart owns the wording, so that a message written by a
   * content provider cannot become the app's own copy.
   */
  private class Refusal(val code: String, val detail: String) : Exception(detail)

  private fun preferences() =
    activity.getSharedPreferences("org.wordbridge.cloud_backup", Context.MODE_PRIVATE)

  /**
   * The chosen folder, or null where there is not one this app may still write
   * to.
   *
   * The permission is checked rather than assumed. A caregiver can revoke it
   * from system settings, and a stored URI that no longer grants anything would
   * otherwise fail on every copy with an error nobody can act on.
   */
  private fun folder(): Uri? {
    val stored = preferences().getString(TREE, null) ?: return null
    val tree = Uri.parse(stored)

    val held =
      activity.contentResolver.persistedUriPermissions.any {
        it.uri == tree && it.isWritePermission
      }
    return if (held) tree else null
  }

  private fun tree(): Uri = folder() ?: throw Refusal("folder", "No folder chosen.")

  /**
   * Where copies go, and whether one would arrive.
   *
   * There is one place on this tablet, so the answer never changes. It is sent
   * anyway because the Dart side asks both platforms the same question — an
   * iPad has two answers, and the shape of the reply is not the place to
   * encode which platform this is.
   *
   * No name is offered. `cloudLabel` already knows what a folder chosen on
   * Android is usually called, and a display name pulled from a content
   * provider would be the provider's copy rather than the app's.
   */
  private fun where(): Map<String, Any> =
    mapOf("place" to "folder", "reachable" to (folder() != null))

  /**
   * Sets up whatever is needed, opening the picker only where one is due.
   *
   * A folder already held is not asked for again. `pick` overrides that, which
   * is how somebody moves to a different provider and the only way back from a
   * folder whose permission the system has dropped.
   */
  private fun connect(call: MethodCall, result: MethodChannel.Result) {
    val pick = call.argument<Boolean>("pick") ?: false
    if (!pick && folder() != null) {
      result.success(where())
      return
    }
    choose(result)
  }

  private fun choose(result: MethodChannel.Result) {
    if (pending != null) {
      result.error("failed", "A folder is already being chosen.", null)
      return
    }
    pending = result

    val intent =
      Intent(Intent.ACTION_OPEN_DOCUMENT_TREE)
        .addFlags(
          Intent.FLAG_GRANT_READ_URI_PERMISSION or
            Intent.FLAG_GRANT_WRITE_URI_PERMISSION or
            Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION
        )
    activity.startActivityForResult(intent, REQUEST)
  }

  /**
   * The picker's answer. Returns whether it was this plugin's.
   *
   * A dismissed picker answers with the state as it stands rather than
   * failing. Somebody who changed their mind has not hit an error, and the Dart
   * side leaves the switch on with a line saying no folder has been chosen
   * yet.
   */
  fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
    if (requestCode != REQUEST) return false

    val result = pending ?: return true
    pending = null

    val tree = data?.data
    if (resultCode != Activity.RESULT_OK || tree == null) {
      result.success(where())
      return true
    }

    // Persisted, or the grant dies with this process and the next automatic
    // copy is a picker in somebody's face.
    activity.contentResolver.takePersistableUriPermission(
      tree,
      Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION
    )
    preferences().edit().putString(TREE, tree.toString()).apply()

    result.success(where())
    return true
  }

  /** What the folder holds. The Dart side decides which of them are snapshots. */
  private fun list(): List<Map<String, Any>> {
    val tree = tree()
    val children =
      DocumentsContract.buildChildDocumentsUriUsingTree(
        tree,
        DocumentsContract.getTreeDocumentId(tree)
      )

    val found = mutableListOf<Map<String, Any>>()
    activity.contentResolver
      .query(
        children,
        arrayOf(
          DocumentsContract.Document.COLUMN_DOCUMENT_ID,
          DocumentsContract.Document.COLUMN_DISPLAY_NAME,
          DocumentsContract.Document.COLUMN_SIZE
        ),
        null,
        null,
        null
      )
      ?.use { cursor ->
        while (cursor.moveToNext()) {
          found.add(
            mapOf(
              "id" to cursor.getString(0),
              "name" to cursor.getString(1),
              "bytes" to cursor.getLong(2).toInt()
            )
          )
        }
      }
    return found
  }

  private fun documentUri(id: String): Uri =
    DocumentsContract.buildDocumentUriUsingTree(tree(), id)

  private fun upload(path: String, name: String): Map<String, Any> {
    val tree = tree()
    val parent =
      DocumentsContract.buildDocumentUriUsingTree(tree, DocumentsContract.getTreeDocumentId(tree))

    // Replaced rather than added beside. Providers happily create a second
    // "wordbridge-….db (1)", and two files claiming one instant is a list of
    // dates a caregiver cannot choose between.
    list().firstOrNull { it["name"] == name }?.let { delete(it["id"] as String) }

    val created =
      DocumentsContract.createDocument(
        activity.contentResolver,
        parent,
        "application/octet-stream",
        name
      )
        ?: throw Refusal("failed", "The folder would not take a new file.")

    val source = java.io.File(path)
    try {
      activity.contentResolver.openOutputStream(created, "wt").use { out ->
        if (out == null) throw Refusal("failed", "The folder would not open for writing.")
        source.inputStream().use { it.copyTo(out) }
      }
    } catch (e: Exception) {
      // Any failure, not only an IO one. Half a database is worse than none:
      // it would be listed, offered and believed. The Dart side checks the
      // length it gets back on the way in as well, but a file that a failed
      // write left behind should not survive the failure that made it.
      DocumentsContract.deleteDocument(activity.contentResolver, created)
      if (e is Refusal) throw e
      throw Refusal(if (outOfSpace(e)) "space" else "offline", e.toString())
    }

    // The name the provider actually used. Some rename on create, and a
    // snapshot's date lives in its name — so the caller is told what it really
    // ended up being called rather than what it asked for.
    val id = DocumentsContract.getDocumentId(created)
    val written = list().firstOrNull { it["id"] == id }

    return mapOf(
      "id" to id,
      "name" to (written?.get("name") ?: name),
      "bytes" to ((written?.get("bytes") as? Int) ?: source.length().toInt())
    )
  }

  private fun download(id: String, path: String) {
    val destination = java.io.File(path)
    try {
      activity.contentResolver.openInputStream(documentUri(id)).use { input ->
        if (input == null) throw Refusal("missing", "No such backup in the folder.")
        destination.outputStream().use { input.copyTo(it) }
      }
    } catch (e: java.io.FileNotFoundException) {
      throw Refusal("missing", e.toString())
    } catch (e: java.io.IOException) {
      destination.delete()
      throw Refusal("offline", e.toString())
    }
  }

  private fun delete(id: String) {
    DocumentsContract.deleteDocument(activity.contentResolver, documentUri(id))
  }

  private fun outOfSpace(e: Exception): Boolean =
    e.message?.contains("space", ignoreCase = true) == true

  private companion object {
    const val TREE = "tree"
    const val REQUEST = 0x8AC4
  }
}

File: app/ios/Runner/CloudBackup.swift
import Flutter
import UIKit

/// Copies of the board in the family's own iCloud, and nowhere else.
///
/// The app's own ubiquity container, under `Documents`. Not the automatic
/// iCloud device backup that sweeps up an app's local documents folder: that
/// one cannot be triggered, listed, dated or verified, and it restores only
/// onto a wiped device. It is the mechanism behind *"we lost months of custom
/// button and phrase building because we thought the iCloud backup would
/// protect us"*, and a backups screen built on it could only ever say "probably".
///
/// Nothing here talks to a server of ours. The destination is the iCloud
/// account already signed in on the device, the entitlement names this app's
/// own container, and no credential of ours exists anywhere in this file.
///
/// Under `Documents` rather than the container root so the files are visible in
/// the Files app: a family moving to a tablet that is not signed in to the same
/// account has to be able to carry the file across by hand. It needs the
/// matching `NSUbiquitousContainers` key in Info.plist.
///
/// Every call runs off the main thread. iCloud coordination blocks, and a
/// blocked main thread on this app is a nonspeaking person staring at a frozen
/// board.
public class CloudBackup: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "org.wordbridge/cloud_backup",
      binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(CloudBackup(), channel: channel)
  }

  /// How long a fetch of one snapshot may wait for iCloud to materialize it.
  ///
  /// Bounded because the caller is a caregiver watching a spinner on the day
  /// their tablet stopped working. A refusal they can retry beats a wait with
  /// no end, and the local backups are still there either way.
  private static let downloadTimeout: TimeInterval = 120

  private let queue = DispatchQueue(label: "org.wordbridge.cloud-backup")

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let arguments = call.arguments as? [String: Any] ?? [:]

    queue.async {
      do {
        let value: Any?
        switch call.method {
        // `connect` has nothing to ask for. The account is the device's, and
        // the container comes with it — the Android side is where somebody has
        // to name a folder.
        case "reachable", "connect":
          value = (try? self.documents()) != nil
        case "list":
          value = try self.list()
        case "upload":
          value = try self.upload(
            path: arguments["path"] as? String ?? "",
            name: arguments["name"] as? String ?? "")
        case "download":
          try self.download(
            id: arguments["id"] as? String ?? "",
            to: arguments["path"] as? String ?? "")
          value = nil
        case "delete":
          try self.delete(id: arguments["id"] as? String ?? "")
          value = nil
        default:
          DispatchQueue.main.async { result(FlutterMethodNotImplemented) }
          return
        }
        DispatchQueue.main.async { result(value) }
      } catch let failure as Refusal {
        DispatchQueue.main.async {
          result(FlutterError(code: failure.code, message: failure.detail, details: nil))
        }
      } catch {
        DispatchQueue.main.async {
          result(FlutterError(code: "failed", message: "\(error)", details: nil))
        }
      }
    }
  }

  /// A reason the Dart side turns into a sentence. The strings never reach a
  /// caregiver — see `refusalFor` in cloud_destination.dart, which owns the
  /// wording so that a platform message cannot become app copy.
  private struct Refusal: Error {
    let code: String
    let detail: String
  }

  /// The container's Documents folder, created if this is the first copy.
  ///
  /// Throws `signIn` rather than returning nil for the ordinary case: an iPad
  /// signed out of iCloud, which is a thing a caregiver can fix and the most
  /// common reason copies stop arriving.
  private func documents() throws -> URL {
    guard FileManager.default.ubiquityIdentityToken != nil,
      let container = FileManager.default.url(forUbiquityContainerIdentifier: nil)
    else {
      throw Refusal(code: "signIn", detail: "No iCloud account on this device.")
    }

    let documents = container.appendingPathComponent("Documents", isDirectory: true)
    if !FileManager.default.fileExists(atPath: documents.path) {
      try FileManager.default.createDirectory(
        at: documents, withIntermediateDirectories: true)
    }
    return documents
  }

  /// What the container holds. Names and sizes only; the Dart side decides
  /// which of them are snapshots.
  ///
  /// `fileSize` first, `totalFileSize` only as a fallback. The second can
  /// include metadata and is a displayable number rather than a byte count;
  /// nothing decides whether a restore is whole from it — see `_wholeDatabase`
  /// in cloud_backup.dart, which opens the file instead — but a caregiver
  /// choosing between dates is shown it, and 0 KB is a list nobody would
  /// trust.
  private func list() throws -> [[String: Any]] {
    let documents = try documents()
    let keys: [URLResourceKey] = [.nameKey, .totalFileSizeKey, .fileSizeKey]

    let entries =
      (try? FileManager.default.contentsOfDirectory(
        at: documents,
        includingPropertiesForKeys: keys,
        options: [.skipsHiddenFiles])) ?? []

    return entries.compactMap { url in
      let values = try? url.resourceValues(forKeys: Set(keys))
      guard let bytes = values?.fileSize ?? values?.totalFileSize else { return nil }
      return ["id": url.lastPathComponent, "name": url.lastPathComponent, "bytes": bytes]
    }
  }

  private func upload(path: String, name: String) throws -> [String: Any] {
    let source = URL(fileURLWithPath: path)
    let destination = try documents().appendingPathComponent(name)

    try coordinate(writing: destination) { target in
      if FileManager.default.fileExists(atPath: target.path) {
        try FileManager.default.removeItem(at: target)
      }
      do {
        try FileManager.default.copyItem(at: source, to: target)
      } catch let error as NSError where error.code == NSFileWriteOutOfSpaceError {
        throw Refusal(code: "space", detail: "No room in iCloud.")
      }
    }

    let bytes =
      (try? destination.resourceValues(forKeys: [.fileSizeKey]).fileSize)
      ?? (try? FileManager.default.attributesOfItem(atPath: destination.path)[.size] as? Int)
      ?? 0

    return ["id": name, "name": name, "bytes": bytes]
  }

  /// Fetches one back, waiting for iCloud to bring it down first.
  ///
  /// A snapshot the device has never opened is a placeholder until something
  /// asks for it, and copying a placeholder produces a file that looks like a
  /// backup and is not one. So the download is started, waited for, and only
  /// then read — and the Dart side opens what it gets anyway, because a wait
  /// that timed out and a file that arrived whole are otherwise
  /// indistinguishable from there.
  private func download(id: String, to path: String) throws {
    let source = try documents().appendingPathComponent(id)
    guard FileManager.default.fileExists(atPath: source.path) else {
      throw Refusal(code: "missing", detail: "No such backup in iCloud.")
    }

    try FileManager.default.startDownloadingUbiquitousItem(at: source)

    let deadline = Date().addingTimeInterval(CloudBackup.downloadTimeout)
    while Date() < deadline {
      let status = try? source.resourceValues(
        forKeys: [.ubiquitousItemDownloadingStatusKey]
      ).ubiquitousItemDownloadingStatus

      if status == .current || status == nil { break }
      Thread.sleep(forTimeInterval: 0.25)
    }

    let destination = URL(fileURLWithPath: path)
    var readError: Error?
    var coordinationError: NSError?

    NSFileCoordinator().coordinate(
      readingItemAt: source, options: [], error: &coordinationError
    ) { readable in
      do {
        if FileManager.default.fileExists(atPath: destination.path) {
          try FileManager.default.removeItem(at: destination)
        }
        try FileManager.default.copyItem(at: readable, to: destination)
      } catch {
        readError = error
      }
    }

    if let error = coordinationError ?? readError {
      throw Refusal(code: "offline", detail: "\(error)")
    }
  }

  private func delete(id: String) throws {
    let target = try documents().appendingPathComponent(id)
    guard FileManager.default.fileExists(atPath: target.path) else { return }

    try coordinate(writing: target) { url in
      try FileManager.default.removeItem(at: url)
    }
  }

  private func coordinate(writing url: URL, _ body: (URL) throws -> Void) throws {
    var thrown: Error?
    var coordinationError: NSError?

    NSFileCoordinator().coordinate(
      writingItemAt: url, options: .forReplacing, error: &coordinationError
    ) { writable in
      do { try body(writable) } catch { thrown = error }
    }

    if let refusal = thrown as? Refusal { throw refusal }
    if let error = thrown ?? coordinationError {
      throw Refusal(code: "offline", detail: "\(error)")
    }
  }
}
